smart-contract/src/contract.rs

```diff
@@ -404,114 +405,132 @@ fn try_redeem(
     // Ensure message sender is a member.
     let member = members_read(deps.storage).load(info.sender.as_bytes())?;
 
+    // Ensure reserve_denom is always the bank reserve token.
+    // Note: originally we supported a specified reserve denom to support other bank reserve tokens.
+    let reserve_denom = reserve_denom.unwrap_or_else(|| member.denom.clone());
+    if reserve_denom != member.denom {
+        return Err(contract_err("unsupported reserve denom"));
+    }
+
     // Ensure member holds at least the requested amount of dcc tokens.
     let balance = deps.querier.query_balance(&member.id, &state.dcc_denom)?;
     if balance.amount < amount {
         return Err(contract_err("insufficient dcc token balance"));
     }
 
-    // Use the member's reserve denom if nothing was requested explicitly.
-    let reserve_denom = reserve_denom.unwrap_or_else(|| member.denom.clone());
+    // Ensure marker holds at least the request amount of reserve token.
+    let querier = ProvenanceQuerier::new(&deps.querier);
+    let member_marker = querier.get_marker_by_denom(&reserve_denom)?;
+    let reserve_balance = member_marker
+        .coins
+        .into_iter()
+        .find(|c| c.denom == reserve_denom)
+        .map(|c| c.amount)
+        .unwrap_or_else(Uint128::zero);
+    if reserve_balance < amount {
+        return Err(contract_err("insufficient reserve token balance"));
+    }
 
-    // Ensure denom is supported by the consortium.
-    if reserve_denom != member.denom {
-        let members = get_members(deps.as_ref())?;
-        let maybe_member = members.iter().find(|m| m.denom == reserve_denom);
-        if maybe_member.is_none() {
-            return Err(contract_err("unsupported reserve denom"));
-        }
+    // Get dcc marker
+    let dcc_marker = querier.get_marker_by_denom(&state.dcc_denom)?;
+
+    // Escrow dcc in the marker account for burn.
+    let res = Response::new()
+        .add_message(transfer_marker_coins(
+            amount.u128(),
+            &state.dcc_denom,
+            dcc_marker.address,
+            info.sender.clone(),
+        )?)
+        // Burn the dcc tokens.
+        .add_message(burn_marker_supply(amount.u128(), &state.dcc_denom)?)
+        // Withdraw reserve token from marker to member.
+        .add_message(withdraw_coins(
+            &reserve_denom,
+            amount.u128(),
+            &reserve_denom,
+            info.sender.clone(),
+        )?)
+        .add_attribute("action", "redeem")
```

Comment:
this isnt necessarily backwards compatible because I remove a bunch of attributes given the simplification of redemption but I think that's fine because no one's really parsing wasm and using the attributes that we know of